src/Sarif.Sarifer/BackgroundAnalyzerBase.cs

```diff
@@ -49,68 +39,93 @@ public abstract class BackgroundAnalyzerBase : IBackgroundAnalyzer
         /// <inheritdoc/>
         public abstract string ToolSemanticVersion { get; }
 
+        private const int DefaultBufferSize = 1024;
+
         /// <inheritdoc/>
-        public async Task AnalyzeAsync(string path, string text, CancellationToken cancellationToken)
+        public async Task<Stream> AnalyzeAsync(string path, string text, CancellationToken cancellationToken)
         {
             text = text ?? throw new ArgumentNullException(nameof(text));
 
             string solutionDirectory = await GetSolutionDirectoryAsync().ConfigureAwait(continueOnCapturedContext: false);
 
-            using (Stream stream = new MemoryStream())
-            using (TextWriter writer = new StreamWriter(stream, Encoding.UTF8))
+            // If we don't have a solutionDirectory, then, we don't need to analyze.
+            if (string.IsNullOrEmpty(solutionDirectory))
             {
-                using (SarifLogger sarifLogger = this.MakeSarifLogger(writer))
+                return Stream.Null;
+            }
+
+            try
+            {
+                var stream = new MemoryStream();
+                using (var writer = new StreamWriter(stream, Encoding.UTF8, DefaultBufferSize, leaveOpen: true))
                 {
-                    sarifLogger.AnalysisStarted();
+                    using (SarifLogger sarifLogger = this.MakeSarifLogger(writer))
+                    {
+                        sarifLogger.AnalysisStarted();
 
-                    // TODO: What do we do when path is null (text buffer with no backing file)?
-                    Uri uri = path != null ? new Uri(path, UriKind.Absolute) : null;
+                        // TODO: What do we do when path is null (text buffer with no backing file)?
+                        Uri uri = path != null ? new Uri(path, UriKind.Absolute) : null;
 
-                    this.AnalyzeCore(uri, text, solutionDirectory, sarifLogger, cancellationToken);
+                        this.AnalyzeCore(uri, text, solutionDirectory, sarifLogger, cancellationToken);
 
-                    sarifLogger.AnalysisStopped(RuntimeConditions.None);
-                }
+                        sarifLogger.AnalysisStopped(RuntimeConditions.None);
+                    }
 
-                await writer.FlushAsync().ConfigureAwait(continueOnCapturedContext: false);
+                    await writer.FlushAsync().ConfigureAwait(continueOnCapturedContext: false);
+                }
 
-                await this.WriteToSinksAsync(path, stream).ConfigureAwait(false);
+                return stream;
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                return Stream.Null;
```

Comment:
>Stream [](start = 23, length = 6)

That's fine, but you still have to dispose the stream :-)

Reply:
it will be disposed by the method that called this.

---
In reply to: [535653423](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535653423) [](ancestors = 535653423)

Reply:
No, the caller will never see this stream, because you returned `null`.

---
In reply to: [535655053](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535655053) [](ancestors = 535655053,535653423)

Reply:
got it.

---
In reply to: [535656794](https://github.com/microsoft/sarif-visualstudio-extension/pull/310#discussion_r535656794) [](ancestors = 535656794,535655053,535653423)